File: bindings/typescript/src/skills.ts
/**
 * RC-SW1 skill-catalog views — a declarative `kortecx.skill/v1` bundle
 * (instructions + a tool grant-WISH set). Kept in its own module (the Rust
 * core's module-per-concern discipline, GR3).
 *
 * SN-8: `skillRef` and `instructionsRef` are SERVER-DERIVED (blake3 over the
 * canonical manifest / the stored body) — the client sends bytes, never an
 * identity. A wish is never authority: attaching a skill grants nothing; at
 * `runApp` the server intersects the wish against the caller's grants and the
 * live broker (`wish ∩ grants ∩ fireable`). The catalog lives in an
 * off-journal `skills.db` sidecar (rebuildable-to-empty), caller-scoped, with
 * UNIFORM not-found. PURE DATA (web-safe).
 */

import type {
  AddSkillResponse as PbAddSkillResponse,
  GetSkillFormResponse as PbGetSkillFormResponse,
  SkillSummary as PbSkillSummary,
} from "./gen/kortecx/v1/gateway_pb.js";
import { encode } from "./hexids.js";

/** The manifest schema/version tag — readers fail closed on a mismatch. */
export const SKILL_SCHEMA = "kortecx.skill/v1";

/** The input to `client.skills.add` — a pack (manifest + body) or a stored-form manifest. */
export interface AddSkillInput {
  /** The `kortecx.skill/v1` manifest object (pack form: NO `instructions_ref`). */
  manifest: Record<string, unknown>;
  /**
   * The instructions markdown body (pack form). Omit iff the manifest already
   * names a 64-hex `instructions_ref` (stored form).
   */
  instructions?: string;
}

/** A stored skill's catalog/display view (the manifest-derived summary + server id). */
export class SkillSummary {
  constructor(
    /** Server-derived canonical-manifest hash, as hex (16 bytes ⇒ 32 hex chars). */
    readonly skillRef: string,
    readonly name: string,
    readonly version: string,
    readonly description: string,
    /** 64-hex content-store ref to the instructions body. */
    readonly instructionsRef: string,
    /** The tool grant-WISH set (id → version); a wish, never a grant. */
    readonly tools: Record<string, string>,
    readonly tags: string[],
  ) {}

  static fromProto(s: PbSkillSummary): SkillSummary {
    return new SkillSummary(
      encode(s.skillRef),
      s.name,
      s.version,
      s.description,
      s.instructionsRef,
      { ...s.tools },
      [...s.tags],
    );
  }
}

/** The outcome of an `AddSkill` upsert (server-derived refs + dedup signal). */
export class AddSkillResult {
  constructor(
    readonly skillRef: string,
    readonly name: string,
    readonly instructionsRef: string,
    readonly deduplicated: boolean,
  ) {}

  static fromProto(r: PbAddSkillResponse): AddSkillResult {
    return new AddSkillResult(encode(r.skillRef), r.name, r.instructionsRef, r.deduplicated);
  }
}

/** One wished tool with the ADVISORY `registered` bit (display only, never a grant). */
export interface SkillWish {
  readonly toolId: string;
  readonly toolVersion: string;
  /** Could THIS serve currently fire it (registered/dialed)? Advisory display. */
  readonly registered: boolean;
}

/** The `GetSkillForm` view: summary + wishes + the instructions preview. */
export class SkillForm {
  constructor(
    readonly summary: SkillSummary,
    readonly wishes: SkillWish[],
    /** Server-capped display excerpt (`''` when the skill was added by ref). */
    readonly instructionsPreview: string,
    readonly previewTruncated: boolean,
  ) {}

  static fromProto(r: PbGetSkillFormResponse): SkillForm | null {
    if (!r.found || !r.summary) return null;
    return new SkillForm(
      SkillSummary.fromProto(r.summary),
      r.wishes.map((w) => ({
        toolId: w.toolId,
        toolVersion: w.toolVersion,
        registered: w.registered,
      })),
      r.instructionsPreview,
      r.previewTruncated,
    );
  }
}

File: bindings/python/src/kortecx/skills.py
"""RC-SW1 skill-catalog views — declarative ``kortecx.skill/v1`` bundles.

A skill is instructions + a tool grant-WISH set; adding one grants NOTHING. At
``run_app`` the server intersects the wish against the caller's grants and the
live broker (``wish ∩ grants ∩ fireable``). SN-8: ``skill_ref`` and
``instructions_ref`` are SERVER-DERIVED — the client sends bytes, never an
identity. The catalog is an off-journal ``skills.db`` sidecar
(rebuildable-to-empty), caller-scoped, with UNIFORM not-found.

Kept in its own module (the Rust core's module-per-concern discipline, GR3).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Dict, List

from . import hexids

if TYPE_CHECKING:  # pragma: no cover - typing only
    from .v1 import gateway_pb2 as _g

SKILL_SCHEMA = "kortecx.skill/v1"
"""The manifest schema/version tag — readers fail closed on a mismatch."""


@dataclass(frozen=True)
class SkillSummary:
    """A stored skill's catalog/display view (manifest-derived + server id)."""

    skill_ref: str  # server-derived canonical-manifest hash, as hex
    name: str
    version: str
    description: str
    instructions_ref: str  # 64-hex content-store ref to the instructions body
    tools: Dict[str, str]  # the tool grant-WISH set (a wish, never a grant)
    tags: List[str]

    @classmethod
    def from_proto(cls, s: "_g.SkillSummary") -> "SkillSummary":
        return cls(
            skill_ref=hexids.encode(s.skill_ref),
            name=s.name,
            version=s.version,
            description=s.description,
            instructions_ref=s.instructions_ref,
            tools=dict(s.tools),
            tags=list(s.tags),
        )


@dataclass(frozen=True)
class AddSkillResult:
    """The outcome of an ``AddSkill`` upsert (server-derived refs + dedup signal)."""

    skill_ref: str
    name: str
    instructions_ref: str
    deduplicated: bool

    @classmethod
    def from_proto(cls, r: "_g.AddSkillResponse") -> "AddSkillResult":
        return cls(
            skill_ref=hexids.encode(r.skill_ref),
            name=r.name,
            instructions_ref=r.instructions_ref,
            deduplicated=r.deduplicated,
        )


@dataclass(frozen=True)
class SkillWish:
    """One wished tool with the ADVISORY ``registered`` bit (display only)."""

    tool_id: str
    tool_version: str
    registered: bool  # could THIS serve currently fire it? never a grant.


@dataclass(frozen=True)
class SkillForm:
    """The ``GetSkillForm`` view: summary + wishes + the instructions preview."""

    summary: SkillSummary
    wishes: List[SkillWish]
    instructions_preview: str  # server-capped excerpt ('' when added by ref)
    preview_truncated: bool

    @classmethod
    def from_proto(cls, r: "_g.GetSkillFormResponse") -> "SkillForm":
        return cls(
            summary=SkillSummary.from_proto(r.summary),
            wishes=[
                SkillWish(
                    tool_id=w.tool_id,
                    tool_version=w.tool_version,
                    registered=w.registered,
                )
                for w in r.wishes
            ],
            instructions_preview=r.instructions_preview,
            preview_truncated=r.preview_truncated,
        )
